src/firewall/firewalld.rs

```diff
@@ -60,19 +69,216 @@ impl firewall::FirewallDriver for FirewallD {
+            Some("org.fedoraproject.FirewallD1.policy"),
+            "getPolicySettings",
+            &(PORTPOLICYNAME),
+        ))?;
+        let policy_config: HashMap<&str, Value> = match policy_config_msg.body() {
+            Ok(m) => m,
+            Err(e) => bail!(
+                "Error decoding DBus message for policy {} configuration: {}",
+                PORTPOLICYNAME,
+                e
+            ),
+        };
+
+        let old_port_forwarding_rules: Array = match policy_config.get("forward_ports") {
+            Some(a) => match a {
+                Value::Array(arr) => arr.clone(),
+                _ => bail!("forward-port in firewalld policy object has a bad type"),
+            },
+            None => {
+                // No existing rules.
+                // Nothing to do - the array must have been wiped already?
+                return Ok(());
+            }
+        };
+
+        let sig = match Signature::try_from("(ssss)") {
+            Ok(s) => s,
+            Err(e) => bail!("Error creating signature for new dbus array: {}", e),
+        };
+        let mut port_forwarding_rules: Array = Array::new(sig);
+        // use an invalid string if we don't have a valid v4 or v6 address.
+        // This is ugly, but easiest code-wise.
+        let ipv4 = match teardown_pf.config.container_ip_v4 {
+            Some(i) => i.to_string(),
+            None => "DOES NOT EXIST".to_string(),
+        };
+        let ipv6 = match teardown_pf.config.container_ip_v6 {
+            Some(i) => i.to_string(),
+            None => "DOES NOT EXIST".to_string(),
+        };
+
+        // Iterate through old rules, remove anything with the IPv4 or IPv6 of
+        // this container as the destination IP.
+        for port_tuple in old_port_forwarding_rules.iter() {
+            match port_tuple {
+                Value::Structure(s) => {
+                    let fields = s.clone().into_fields();
+                    if fields.len() != 4 {
+                        bail!("Port forwarding rule that was not a 4-tuple encountered");
+                    }
+                    let port_ip = match fields[3].clone() {
+			Value::Str(s) => s.as_str().to_string(),
+			_ => bail!("Port forwarding tuples must contain only strings, encountered a non-string object"),
+		    };
+                    debug!("IP string from firewalld is {}", port_ip);
+                    if port_ip != ipv4 && port_ip != ipv6 {
+                        port_forwarding_rules.append(port_tuple.clone())?;
+                    }
+                }
+                _ => bail!("Port forwarding rule that was not a structure encountered"),
+            }
+        }
+
+        let new_rules = Value::new(port_forwarding_rules);
+        let mut new_policy_config = HashMap::<&str, &Value>::new();
+        new_policy_config.insert("forward_ports", &new_rules);
+
+        // Send the updated configuration back to firewalld.
+        match block_on(self.conn.call_method(
```

Comment:
Is there any specific reason to block if api is `async` in nature ? I don't think its a blocker though. Just want to  confirm.

Reply:
There's no reason for us *to* be async. Everything has to happen in sequence, else we'll be interacting with firewalld objects that don't exist yet. Blocking requests make the most sense for this.

Reply:
Fair.